dist-tracing/src/trace.rs

```diff
@@ -0,0 +1,255 @@
+                    let res = registry
+                        .span(target_id)
+                        .expect("span data not found during eval_ctx for current_trace_ctx");
+                    *st = res.parent().map(|x| x.id());
+                    Some(res)
+                }
+                None => None,
+            });
+
+            trace_ctx_registry
+                .eval_ctx(iter)
+                .map(|x| TraceCtx {
+                    parent_span: Some(trace_ctx_registry.span_id(current_span_id.clone())),
+                    trace_id: x.trace_id,
+                })
+                .ok_or(TraceCtxError::NoParentNodeHasTraceCtx)
+        }
+
+        let span = tracing::Span::current();
+        let res = span
+            .with_subscriber(inner)
+            .ok_or(TraceCtxError::NoEnabledSpan)?;
+        res
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct SpanId {
+    pub tracing_id: tracing::Id,
+    pub instance_id: u64,
+}
+
+// TODO: round trip property test for this
+impl SpanId {
+    /// attempts to parse a SpanId from a string
+    pub fn from_string(s: &str) -> Option<SpanId> {
+        let mut iter = s.split('-');
+        let s1 = iter.next()?;
+        let u1 = u64::from_str_radix(s1, 10).ok()?;
+        let s2 = iter.next()?;
+        let u2 = u64::from_str_radix(s2, 10).ok()?;
+
+        Some(SpanId {
+            tracing_id: tracing::Id::from_u64(u1),
+            instance_id: u2,
+        })
+    }
+
+    /// convert a SpanId to a string that can be included in headers, RPC framework metadata fields, etc
+    pub fn to_string(&self) -> String {
+        format!("{}", self)
+    }
+}
+
+impl std::fmt::Display for SpanId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}-{}", self.tracing_id.into_u64(), self.instance_id)
+    }
+}
+
+/// NOTE: why not just have this be a u128? It'd be so much better...
+/// A Honeycomb Trace ID. Uniquely identifies a single distributed (potentially multi-process) trace.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct TraceId(pub(crate) u128);
+
+impl TraceId {
+    /// attempts to parse a TraceId from a string
+    pub fn from_string(s: &str) -> Option<Self> {
+        let u = u128::from_str_radix(s, 10).ok()?;
+        Some(TraceId(u))
+    }
+
+    /// convert a TraceId to a string that can be included in headers, RPC framework metadata fields, etc
+    pub fn to_string(&self) -> String {
+        format!("{}", self.0)
+    }
+
+    /// Generate a random trace ID by using a thread-level RNG to generate a u128
+    pub fn generate() -> Self {
+        use rand::Rng;
+        let u: u128 = rand::thread_rng().gen();
+
+        TraceId(u)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Span<'a, V> {
+    pub id: SpanId,
+    pub trace_id: TraceId,
+    pub parent_id: Option<SpanId>,
+    pub initialized_at: DateTime<Utc>,
+    pub elapsed_ms: i64,
```

Comment:
why is this signed? is there a scenario where a span would take a negative amount of time? also, seems like this should just be a `Duration`...